File: python-v1-archive/storage.py
import json
import os
from datetime import datetime
from werkzeug.security import generate_password_hash, check_password_hash

LOG_FILE = "logs.json"
VENDORS_FILE = "vendors.json"
USERS_FILE = "users.json"

# --- User Management ---

def get_users():
    if not os.path.exists(USERS_FILE):
        return []
    with open(USERS_FILE, "r") as f:
        try:
            return json.load(f).get("users", [])
        except json.JSONDecodeError:
            return []

def save_users(users):
    with open(USERS_FILE, "w") as f:
        json.dump({"users": users}, f, indent=2)

def add_user(username, password):
    users = get_users()
    if any(u['username'] == username for u in users):
        return False, "Username already exists"
    
    password_hash = generate_password_hash(password)
    users.append({
        "username": username,
        "password_hash": password_hash
    })
    save_users(users)
    return True, "User created successfully"

def verify_user(username, password):
    users = get_users()
    user = next((u for u in users if u['username'] == username), None)
    if user and check_password_hash(user['password_hash'], password):
        return True
    return False

# --- Log Management (Scoped) ---

def save_log(username, phone, name, vendor, status):
    """Saves a log entry scoped to a user."""
    log_entry = {
        "username": username,
        "phone": phone,
        "name": name,
        "vendor": vendor,
        "status": status,
        "timestamp": datetime.now().isoformat()
    }

    if not os.path.exists(LOG_FILE):
        data = {"logs": []}
    else:
        with open(LOG_FILE, "r") as f:
            try:
                data = json.load(f)
            except json.JSONDecodeError:
                data = {"logs": []}

    data["logs"].append(log_entry)
    with open(LOG_FILE, "w") as f:
        json.dump(data, f, indent=2)

def get_logs(username):
    """Retrieves logs for a specific user."""
    if not os.path.exists(LOG_FILE):
        return []
    with open(LOG_FILE, "r") as f:
        try:
            data = json.load(f)
            logs = data.get("logs", [])
            return [l for l in logs if l.get("username") == username]
        except json.JSONDecodeError:
            return []

# --- Vendor Management (Scoped) ---

def get_vendors(username):
    """Retrieves vendors for a specific user."""
    if not os.path.exists(VENDORS_FILE):
        return []
    with open(VENDORS_FILE, "r") as f:
        try:
            data = json.load(f)
            vendors = data.get("vendors", [])
            return [v for v in vendors if v.get("username") == username]
        except json.JSONDecodeError:
            return []

def save_vendors(vendors):
    """Saves the entire list of vendors."""
    with open(VENDORS_FILE, "w") as f:
        json.dump({"vendors": vendors}, f, indent=2)

def add_vendor(username, name, maps_link):
    """Adds a new vendor scoped to a user."""
    # Load ALL vendors to maintain the full list
    if not os.path.exists(VENDORS_FILE):
        all_vendors = []
    else:
        with open(VENDORS_FILE, "r") as f:
            try:
                all_vendors = json.load(f).get("vendors", [])
            except json.JSONDecodeError:
                all_vendors = []

    vendor_id = f"v{len(all_vendors) + 1}_{int(datetime.now().timestamp())}"
    new_vendor = {
        "id": vendor_id,
        "username": username,
        "name": name,
        "maps_link": maps_link
    }
    all_vendors.append(new_vendor)
    save_vendors(all_vendors)
    return new_vendor

def delete_vendor(username, vendor_id):
    """Removes a vendor if it belongs to the user."""
    if not os.path.exists(VENDORS_FILE):
        return
    
    with open(VENDORS_FILE, "r") as f:
        try:
            all_vendors = json.load(f).get("vendors", [])
        except json.JSONDecodeError:
            return

    # Filter out the vendor only if it matches ID AND username
    all_vendors = [v for v in all_vendors if not (v["id"] == vendor_id and v["username"] == username)]
    save_vendors(all_vendors)

File: lib/storage.ts
import bcrypt from 'bcryptjs';
import { supabase } from './supabase';

// Types
export interface User {
  username: string;
  passwordHash: string;
}

export interface Vendor {
  id: string;
  username: string; // scoped to user
  name: string;
  maps_link: string;
}

export interface ActivityLog {
  id: string;
  username: string;
  phone: string;
  name: string;
  vendor: string;
  status: 'opened' | 'failed';
  timestamp: string;
}

// --- User Management ---

export async function addUser(username: string, password: string): Promise<{ success: boolean; message: string }> {
  const { data: existingUser } = await supabase
    .from('users')
    .select('username')
    .eq('username', username)
    .single();
  
  if (existingUser) {
    return { success: false, message: "Username already exists" };
  }

  const passwordHash = await bcrypt.hash(password, 10);
  
  const { error } = await supabase
    .from('users')
    .insert([{ username, password_hash: passwordHash }]);

  if (error) {
    return { success: false, message: "Failed to register user in database" };
  }

  return { success: true, message: "User registered successfully" };
}

export async function verifyUser(username: string, password: string): Promise<User | null> {
  const { data: user, error } = await supabase
    .from('users')
    .select('username, password_hash')
    .eq('username', username)
    .single();

  if (error || !user) return null;

  const hash = user.password_hash;
  if (!hash) return null;

  // Check if it's a bcrypt hash
  if (hash.startsWith('$2b$') || hash.startsWith('$2a$')) {
    if (await bcrypt.compare(password, hash)) {
      return {
        username: user.username,
        passwordHash: user.password_hash
      } as User;
    }
  }
  
  return null;
}

// --- Vendor Management ---

export async function getVendors(username: string): Promise<Vendor[]> {
  const { data, error } = await supabase
    .from('vendors')
    .select('*')
    .eq('username', username);

  if (error) return [];
  return data as Vendor[];
}

export async function addVendor(username: string, name: string, maps_link: string): Promise<Vendor | null> {
  const newVendor = {
    id: Math.random().toString(36).substring(2, 9),
    username,
    name,
    maps_link
  };

  const { error } = await supabase
    .from('vendors')
    .insert([newVendor]);

  if (error) return null;
  return newVendor;
}

export async function deleteVendor(username: string, vendorId: string): Promise<void> {
  await supabase
    .from('vendors')
    .delete()
    .eq('id', vendorId)
    .eq('username', username);
}

// --- Logging ---

export async function getLogs(username: string): Promise<ActivityLog[]> {
  const { data, error } = await supabase
    .from('activity_logs')
    .select('*')
    .eq('username', username)
    .order('timestamp', { ascending: false });

  if (error) return [];
  return data as ActivityLog[];
}

export async function saveLog(username: string, phone: string, name: string, vendor: string, status: 'opened' | 'failed'): Promise<ActivityLog | null> {
  const newLog = {
    id: Math.random().toString(36).substring(2, 9),
    username,
    phone,
    name,
    vendor,
    status,
    timestamp: new Date().toISOString()
  };

  const { error } = await supabase
    .from('activity_logs')
    .insert([newLog]);

  if (error) return null;
  return newLog as ActivityLog;
}
